x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java

```diff
@@ -371,7 +375,7 @@ private void sendResponseOrFailure(String jobId,
                                                                AtomicArray<Exception> failures) {
                                 List<Exception> caughtExceptions = failures.asList();
                                 if (caughtExceptions.size() == 0) {
-                                    listener.onResponse(new CloseJobAction.Response(true));
+                                    closeRunningForecasts(jobId, listener);
```

Comment:
Isn't `sendRemoveRequest` async? What if other results are still in the results queue and overwrite the failure state again?

I wonder if it would be better to do this in `AutodetectCommunicator::killProcess`?

Reply:
🤔 Good point. 

But, `awaitCompletion` is false when the task is cancelled. We might could throw a hook in results processor somewhere, but I am not sure where or if it is worth it.

I am unsure where to draw the line of "good enough" and "we absolutely don't want opened forecast docs sitting around".

I will keep investigating along these lines to see if there is a natural place for a "execute this action after you are done processing results" type of thing.

Reply:
I added a commit doing exactly this. Going to update tests, etc.. changing PR to draft while I iron out the wrinkles.